### PR TITLE
BouncyCastle support

### PR DESCRIPTION
**Pull request by koh-osug**
Motivation:

I experienced an [issue](https://github.com/chhsiao90/nitmproxy/issues/7) under Android not being able to load a web page. It might be related to ALPN of the internal TLS implementation. I was trying BouncyCastle as a replacement but this was not supported by Netty. So I added this.
Also BouncyCastle is a pure Java solution, all the other providers (OpenSSL, Conscrypt) require native libraries which are not available under Android at least.

Modification:

I added a `BouncyCastleAlpnSslEngine.java` and support classes. It is relying on the JDK code, hence I had to open some classes to prevent code duplication.

Maybe you can point be to some test cases to reuse.

Result:

BouncyCastle can be used as TLS provider with:

~~~java
                Security.insertProviderAt(new BouncyCastleJsseProvider(), 1);
                Security.addProvider(new BouncyCastleProvider());
                Security.addProvider(new SecureRandomProvider());
~~~

SecureRandomProvider is a necessary fix for BouncyCastle otherwise a `Caused by: java.lang.IllegalStateException: unable to create JcaTlsCrypto: DEFAULT SecureRandom not available` exception is thrown. This problem is BC related.

~~~java
/**
 * A JCE security provider for fixing: Caused by: java.lang.IllegalStateException: unable to create JcaTlsCrypto: DEFAULT SecureRandom not available for BouncyCastle when using bctls.
 */
public class SecureRandomProvider extends Provider {
    public static final String PROVIDER_NAME = "SecureRandom";

    public SecureRandomProvider() {
        super(PROVIDER_NAME, 1.0, "SecureRandom provider.");

        putService(new Service(this, "SecureRandom", "DEFAULT",
                SecureRandom.class.getName(), null, null));

    }

    public static class SecureRandom extends SecureRandomSpi {

        private final java.security.SecureRandom SECURE_RANDOM = new java.security.SecureRandom();

        @Override
        protected void engineSetSeed(byte[] seed) {
            SECURE_RANDOM.setSeed(seed);
        }

        @Override
        protected void engineNextBytes(byte[] bytes) {
            SECURE_RANDOM.nextBytes(bytes);
        }

        @Override
        protected byte[] engineGenerateSeed(int numBytes) {
            return SECURE_RANDOM.generateSeed(numBytes);
        }
    }
}
~~~

### CODE REVIEW

File: handler/src/main/java/io/netty/handler/ssl/BouncyCastleAlpnSslEngine.java [hunk 1]
```diff
@@ -0,0 +1,66 @@
+/*
+ * Copyright 2017 The Netty Project
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/BouncyCastleAlpnSslEngine.java:2]: 2021

File: handler/src/main/java/io/netty/handler/ssl/BouncyCastle.java [hunk 2]
```diff
@@ -0,0 +1,52 @@
...
+package io.netty.handler.ssl;
+
+import javax.net.ssl.SSLEngine;
+
+/**
+ * Contains methods that can be used to detect if BouncyCastle is usable.
+ */
+final class BouncyCastle {
+
+    private static boolean IS_BC_SSLENGINE;
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/BouncyCastle.java:25]: rename to `BOUNCY_CASTLE_ON_CLASSPATH` or something like that ?

Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/BouncyCastle.java:25]: final

File: handler/src/main/java/io/netty/handler/ssl/BouncyCastleAlpnSslUtils.java [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.lang.reflect.Proxy;
+import java.security.AccessController;
+import java.security.PrivilegedExceptionAction;
+import java.util.List;
+import java.util.function.BiFunction;
+
+@SuppressJava6Requirement(reason = "Usage guarded by java version check")
+final class BouncyCastleAlpnSslUtils extends JdkAlpnSslUtils {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/BouncyCastleAlpnSslUtils.java:37]: Not need to extend `JdkAlpnSslUtils` ... this also means `JdkAlpnSslUtils` should stay final

File: handler/src/main/java/io/netty/handler/ssl/JdkAlpnSslUtils.java [hunk 4]
```diff
@@ -32,7 +32,7 @@
 import io.netty.util.internal.logging.InternalLoggerFactory;
 
 @SuppressJava6Requirement(reason = "Usage guarded by java version check")
-final class JdkAlpnSslUtils {
+class JdkAlpnSslUtils {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/JdkAlpnSslUtils.java:35]: please revert changes in this class

**Review by normanmaurer — CHANGES_REQUESTED** [review]
Also please sign our ICLA if you didn't do yet: 

https://netty.io/s/icla

File: handler/src/main/java/io/netty/handler/ssl/JdkAlpnSslEngine.java [hunk 5]
```diff
@@ -15,28 +15,35 @@
  */
 package io.netty.handler.ssl;
 
+import io.netty.buffer.ByteBuf;
+import io.netty.buffer.ByteBufUtil;
 import io.netty.util.internal.StringUtil;
 import io.netty.util.internal.SuppressJava6Requirement;
+import io.netty.util.internal.logging.InternalLogger;
+import io.netty.util.internal.logging.InternalLoggerFactory;
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/JdkAlpnSslEngine.java:23]: revert the imports that are not used...

File: handler/src/main/java/io/netty/handler/ssl/JdkAlpnSslEngine.java [hunk 6]
```diff
@@ -15,28 +15,35 @@
...
+import java.util.function.BiConsumer;
 import java.util.function.BiFunction;
 
 import static io.netty.handler.ssl.SslUtils.toSSLHandshakeException;
 import static io.netty.handler.ssl.JdkApplicationProtocolNegotiator.ProtocolSelectionListener;
 import static io.netty.handler.ssl.JdkApplicationProtocolNegotiator.ProtocolSelector;
 
 @SuppressJava6Requirement(reason = "Usage guarded by java version check")
-final class JdkAlpnSslEngine extends JdkSslEngine {
+class JdkAlpnSslEngine extends JdkSslEngine {
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/JdkAlpnSslEngine.java:42]: revert

**Review by normanmaurer — CHANGES_REQUESTED** [review]

File: handler/src/main/java/io/netty/handler/ssl/JdkAlpnSslEngine.java (hunk 6 above)
Comment (koh-osug) [reply handler/src/main/java/io/netty/handler/ssl/JdkAlpnSslEngine.java:36]: This does not work. BouncyCastleAlpnSslEngine is extending the JdkAlpnSslEngine. Not dping this means duplicating almost a whole identical class. I have resolved the other comments.

File: handler/src/main/java/io/netty/handler/ssl/JdkAlpnApplicationProtocolNegotiator.java [hunk 7]
```diff
@@ -16,6 +16,7 @@
 package io.netty.handler.ssl;
 
 import io.netty.buffer.ByteBufAllocator;
+import io.netty.util.internal.PlatformDependent;
```
Comment (normanmaurer) [inline handler/src/main/java/io/netty/handler/ssl/JdkAlpnApplicationProtocolNegotiator.java:19]:
```suggestion
```

**Comment by normanmaurer** [comment]
@koh-osug thanks a lot !

**Comment by koh-osug** [comment]
My pleasure.

**Pull request merged**